lua/starfall/libs_cl/vgui.lua

```diff
@@ -0,0 +1,2782 @@
+local dmen_methods, dmen_meta, dmenwrap, dmenunwrap = instance.Types.UIMenu.Methods, instance.Types.UIMenu, instance.Types.UIMenu.Wrap, instance.Types.UIMenu.Unwrap
+local dmeno_methods, dmeno_meta, dmenowrap, dmenounwrap = instance.Types.UIMenuOption.Methods, instance.Types.UIMenuOption, instance.Types.UIMenuOption.Wrap, instance.Types.UIMenuOption.Unwrap
+local col_meta, cwrap, cunwrap = instance.Types.Color, instance.Types.Color.Wrap, instance.Types.Color.Unwrap
+local plyunwrap = instance.Types.Player.Unwrap
+local vgui_library = instance.Libraries.vgui
+
+function pnl_meta:__tostring()
+	return "UIPanelBase"
+end
+
+function dpnl_meta:__tostring()
+	return "UIPanel"
+end
+
+function dfrm_meta:__tostring()
+	return "UIFrame"
+end
+
+function dscrl_meta:__tostring()
+	return "UIScrollPanel"
+end
+
+function dbut_meta:__tostring()
+	return "UIButton"
+end
+
+function dlab_meta:__tostring()
+	return "UILabel"
+end
+
+function aimg_meta:__tostring()
+	return "UIAvatarImage"
+end
+
+function dprg_meta:__tostring()
+	return "UIProgress"
+end
+
+function dtxe_meta:__tostring()
+	return "UITextEntry"
+end
+
+function dimg_meta:__tostring()
+	return "UIImage"
+end
+
+function dimgb_meta:__tostring()
+	return "UIImageButton"
+end
+
+function dchk_meta:__tostring()
+	return "UICheckBox"
+end
+
+function dnms_meta:__tostring()
+	return "UINumSlider"
+end
+
+function dcom_meta:__tostring()
+	return "UIComboBox"
+end
+
+function dclm_meta:__tostring()
+	return "UIColorMixer"
+end
+
+function dfb_meta:__tostring()
+	return "UIFileBrowser"
+end
+
+function dmen_meta:__tostring()
+	return "UIMenu"
+end
+
+function dmeno_meta:__tostring()
+	return "UIMenuOption"
+end
+
+local function unwrap(pnl)
+	local pnc = tostring(pnl)
+	local unwrapFunc = instance.Types[pnc].Unwrap
+	
+	return unwrapFunc(pnl)
+end
+
+--- Returns how many more VGUI panels the chip can create.
+-- @return number Panels left.
+function vgui_library.panelsLeft()
+	if not SF.Permissions.hasAccess(instance,  nil, "vgui.create") then return 0 end
+	return panelCount:check(instance.player)
+end
+
+--- Returns the panel the cursor is hovering above.
+--- This returns a cached value that is only updated after rendering and before the next VGUI Think/Layout pass.
+-- @return any The hovered panel.
+function vgui_library.getHoveredPanel()
+	return pnlwrap(vgui.GetHoveredPanel())
+end
+
+--- Returns the panel which is currently receiving keyboard input.
+-- @return any The panel with keyboard focus.
+function vgui_library.getKeyboardFocus()
+	return pnlwrap(vgui.GetKeyboardFocus())
+end
+
+--- Sets the position of the panel's top left corner.
+--@param number x The x coordinate of the position.
+--@param number y The y coordinate of the position.
+function pnl_methods:setPos(x, y)
+	checkluatype(x, TYPE_NUMBER)
+	checkluatype(y, TYPE_NUMBER)
+	local uwp = unwrap(self)
+	
+	uwp:SetPos(x, y)

```

Comment:
As my separate comment said, could slim down this code a bunch if the unwraps were inlined.

Reply:
Done